File: 作业2/work02/code/task.cpp
#include <iostream>
//导入OpenCV库
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

int main()
{
    ////任务一：使用OpenCV读取一张测试图片
    //定义图片路径
    string image_name = "test.jpg";
    string image_path = "/home/administrator/myproj/work02/photo/" + image_name;
    Mat image = imread(image_path);

    //判断图片是否读取成功
    if (image.empty())
    {
        cout << "无图片，图片读取失败" << endl;
        return -1;
    }
    cout << "图片读取成功" << endl;

    ////任务二：输出图片基本信息
    int height = image.rows;
    int width = image.cols;
    int channels = image.channels();
    string pixel_type = "uint8 (CV_8U)";

    cout << "图片尺寸(宽度×高度): " << width << " × " << height << endl;
    cout << "图像通道数：" << channels << endl;
    cout << "像素数据类型：" << pixel_type << endl;

    //任务三：显示原图
    namedWindow("原图", WINDOW_NORMAL);//建立一个图片窗口，名字为“原图”
    resizeWindow("原图", 800, 600);//强制窗口大小
    imshow("原图", image);//显示原图
    cout << "请查看窗口，按任意键继续..." << endl;
    waitKey(0);
    destroyWindow("原图");//关闭窗口

    //任务四：灰度图并显示
    Mat image_gray;
    cvtColor(image, image_gray, COLOR_BGR2GRAY); 
    namedWindow("灰度图", WINDOW_NORMAL);//建立一个图片窗口，名字为“灰度图”
    resizeWindow("灰度图", 800, 600);//强制窗口大小
    imshow("灰度图", image_gray);//显示灰度图
    cout << "请查看窗口，按任意键继续..." << endl;
    waitKey(0);
    destroyWindow("灰度图");//关闭窗口

    //任务五：保存灰度图
    string gray_save_name = "test_gray.jpg";
    string gray_save_path = "/home/administrator/myproj/work02/photo/" + gray_save_name;
    imwrite(gray_save_path, image_gray);
    cout << "灰度图保存成功，路径：" << gray_save_path << endl;

    //任务六：像素操作
    Vec3b pixel = image.at<Vec3b>(50, 50);
    cout << "位置(50,50)的像素值: " 
         << (int)pixel[0] << " " 
         << (int)pixel[1] << " " 
         << (int)pixel[2] << endl;

    //修改一块像素为绿色
    Mat image_copy = image.clone();
    for (int y = 15; y < 25; y++)
    {
        for (int x = 43; x < 53; x++)
        {
            image_copy.at<Vec3b>(y, x) = Vec3b(0, 255, 0);
        }
    }

    string pixel_change_save_name = "pixel_change.jpg";
    string pixel_change_save_path = "/home/administrator/myproj/work02/photo/" + pixel_change_save_name;
    imwrite(pixel_change_save_path, image_copy);
    cout << "像素修改完成，保存路径：" << pixel_change_save_path << endl;

    return 0;
}

File: code/作业1/code/task.py
####任务一：使用OpenCV读取一张测试图片
#导入OpenCV库
import cv2
import matplotlib.pyplot as plt

#定义图片路径
image_name="test.jpg"
image_path=f"work01/images/{image_name}"

#通过OpenCV的imread的功能，根据定义的图片路径读取图片数据，并将读取的图片数据存到image里面
image=cv2.imread(image_path)

#判断图片是否读取成功
if image is None:
    print("无图片，图片读取失败")
else:
    print("图片读取成功")
    ####任务二：输出图片基本信息
    #图片的尺寸：高度，宽度
    height,width=image.shape[:2]
    #图片通道数：三个通道说明为彩色图,一个通道说明是灰度图
    if len(image.shape)==3:
        channels=image.shape[2]
    else:
        channels=1
    #像素数据类型
    pixel_type=image.dtype

    #在终端打印
    print(f"图片尺寸(宽度×高度):{height} × {width}")
    print(f"图像通道数：{channels}")
    print(f"像素数据类型：{pixel_type}")

    ####任务三：显示原图：用Matplotlib显示图片
    #排列图片，子图一
    plt.subplot(1,2,1)
    #将OpenCV读取的BGR格式转化为Matplotlib的RGB格式
    image_rgb=cv2.cvtColor(image,cv2.COLOR_BGR2RGB)
    #图片显示
    plt.imshow(image_rgb)
    #隐藏坐标轴
    plt.axis("off")
    #图片标题
    plt.title("picture")

    ####任务四：转换为灰度图并显示
    #排列图片，子图二
    plt.subplot(1,2,2)
    #将彩色图转换为灰度图
    image_gray=cv2.cvtColor(image,cv2.COLOR_BGR2GRAY)
    #用Matplotlib显示灰度图
    plt.imshow(image_gray,cmap='gray')
    plt.axis("off")
    plt.title("grayscale image")

    #图片显示
    plt.show()

    ####任务五：保存灰度图
    if image_gray is not None:
        #保存灰度图的文件名
        gray_save_name="test_gray.jpg"
        #保存灰度图的路径
        gray_save_path=f"work01/images/{gray_save_name}"

        #通过OpenCV的imwrite函数保存图片
        cv2.imwrite(gray_save_path,image_gray)
        print(f"灰度图保存成功，保存路径为：{gray_save_path}")
    else:
        print("灰度图生成失败，无法保存")

    ####任务六：用NumPY做操作
    #输出(50,50)位置的像素值
    pixel=image[50,50]
    print(f"位置(50,50)的像素值:{pixel}")

    #修改某一块像素值
    #复制原图（不修改原图）
    image_copy=image.copy()
    #把（15:25, 43:53）位置的颜色改成鲜绿色
    image_copy[15:25, 43:53]=[0,255,0]
    #定义保存路径
    pixel_change_save_name="pixel_change.jpg"
    pixel_change_save_path=f"work01/images/{pixel_change_save_name}"
    cv2.imwrite(pixel_change_save_path,image_copy)
    print(f"像素修改完成，保存路径为：{pixel_change_save_path}")
